design/20210302.certificate-request-policy.md

```diff
@@ -0,0 +1,724 @@
+---
+title: CertificateRequest Policy
+authors:
+  - "@joshvanl"
+reviewers:
+  - "@cert-manager-dev"
+approvers:
+  - "@cert-manager-dev"
+editor: "@joshvanl"
+creation-date: 2021-03-02
+last-updated: 2021-03-09
+status: provisional
+---
+
+# CertificateRequest Policy
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  * [API](#api)
+  * [Behaviour](#behaviour)
+  * [RBAC](#rbac)
+- [External Policy Server](#external-policy-server)
+  * [API](#api-1)
+  * [gRPC](#grpc)
+- [User Stories](#user-stories)
+  * [Case 1](#case-1)
+  * [Case 2](#case-2)
+  * [Case 4](#case-4)
+  * [Case 5](#case-5)
+  * [Case 6](#case-6)
+<!-- /toc -->
+
+## Summary
+
+Cluster operators will create and manage a set of policy resources that describe
+what and how certificates can be signed, as well as RBAC to manage who can
+request these certificates. Consumers will be able to request certificates,
+restricted by the policy they are bound to.
+
+
+## Motivation
+
+Policy is an important mechanism for security focused users who would like to
+restrict what, who and how certificates are signed in Kubernetes. Currently, the
+only mechanism for restriction is either via namespacing (via namespace scoped
+Issuer resources) where consumers are bucketed between namespaces, or relying on
+policy that is enforced by the signing Issuer.
+
+This greatly limits the amount of restriction that can be imposed by forcing use
+of certain Issuers, limited protection through namespaing, and no scope for
+users to use external bespoke policy enforcement.
+
+
+## Goals
+- Create first class support for policy over CertificateRequests
+- Allow for Policy to be extended to certificates.k8s.io if cert-manager gains adoption of this API
+- Implement a mechanism for allowing pluggable extensible policy
+- Policy management should have similar user experience to existing Kubernetes
+  resources (network policy, PSP replacement)
+
```

Comment:
A native policy resource does offer some benefits, but have we considered using policy engines like OPA/Gatekeeper and Kyverno? 

Seems to me, that a policy engine may offer more overall flexibility to administrators, and perhaps a better UX to developers, as violations can be blocked during admission controls.

However, policies may be harder to write and maintain. 

I am a Kyverno maintainer, and based on interest would be happy to explore how the use cases mentioned can be addressed by Kyverno, and evaluating tradeoffs across approaches.

Reply:
Yes, we would be interested in that. This native policy resource is definitely not meant to replace something like Kyverno, but may be a simple way to solve some common problems. Our idea is to make the interface extensible so that there can be integrations for OPA, Kyverno and any other solution that someone would want to use for policy. We are currently discussing what that interface should be, so your input would be valuable there. Once we have an interface it would be great to discuss building a Kyverno implementation of the interface.

Reply:
I do see the value in a native policy object, and the tradeoffs are interesting.

Kyverno runs as a mutating and validating webhook and applies policies during admission controls and (optionally) as background scans. So, I don't see how it, or other admission controllers, like OPA/Gatekeeper would fit in as an policy integration.


Here is what a sample Kyverno policy would look like - this enforces that a duration is defined and less than 90 days:

```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata: 
  name: certificate-request-policies
spec: 
  validationFailureAction: enforce
  rules: 
  - name: validate-duration
    match: 
      resources: 
        kinds: 
        - cert-manager.io/v1/CertificateRequest
    validate: 
      message: "duration cannot exceed 90 days"
      pattern: 
        spec: 
          duration: "<=90d"
```

Reply:
With the Kyverno policy above, requests that violate the policy will be rejected by the API server as `validationFailureAction` is set to `enforce`.  Setting that to `audit` will allow the request, and report a policy error.

Reply:
Yep, there is definitely room for people to use Kyverno/Gatekeeper webhooks to enforce some policy.

@JoshVanL would you say something (and maybe add to the doc?) about why validating webhooks don't cover all the use cases in cert-manager and why we are proposing this approver model please?

Reply:
Hi @JimBugwadia, there are a few reasons why I am proposing taking this approach, rather than relying on bespoke admission webhooks;

1. The approval condition mirrors the same behaviour as the [upstream certificate.k8s.io CSR type](https://github.com/kubernetes/api/blob/51a1c5553d6812f2f7994f1ca31ddae1d2bcb642/certificates/v1/types.go#L153). Since [adopting this resource](https://github.com/jetstack/cert-manager/pull/3646) in cert-manager is something the project is looking to do, it makes sense to align with it as closely as possible to make migration manageable.
2. Having the resource blocked at admission time doesn't cover all use cases, such as manual approval detailed in the doc. Some of these kinds of policy use cases that we are trying to cover are, at least to me, incompatible with resources being blocked at admission with very limited timeouts.
3. The policy designed here is intended to be extensible. By having an "ExternalPolicyServer", we can create a shim to whatever external policy server we would like. If this design goes ahead, I fully expect cert-manager to create and maintain a number of these shims that allow for things such as OPA or Kyverno to be plugged in, on top of cert-manager base policy.
4. Having a single resource to manage the approval of cert-manager means there is a single pane of glass when it comes to an operator defining what policy is assigned to what through RBAC. An operator can then be confident which policy is applied to which users, and external tooling can quite easily discover this mapping giving the operator confidence policy is setup correctly
5. Some policy decisions need to be turning complete and completely bespoke to the use case. I usually use the example of "certificates can only be signed on Tuesdays", but a real use case is for example istio- here certs need URIs that match the requestor in the form `spiffe://<sa>.<namespace><trust domain>`. This is where external policy again comes in, where operators can write arbitrary code that is enforced using the same RBAC mapping.
6. Users can still install admission webhooks, and this is possible today with or without this implementation. The benefit of creating a base policy resource is that users can start with policy straight away, without needing to install extra components.

Hopefully this helps explains my thought process with the design :smile: